operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -0,0 +1,635 @@
+}
+
+// InsertCEPInCache is used to insert CEP in local cache, this may result in creating a new
+// CES object or updating an existing CES object.
+func (c *cesMgr) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string) string {
+	log.WithFields(logrus.Fields{
+		"CEPName": GetCEPNameFromCCEP(cep, ns),
+	}).Debug("Insert CEP in local cache")
+
+	// check the given cep is already exists in any of the CES.
+	// if yes, Update a ces with the given cep object.
+	if cesName, exists := c.desiredCESs.getCESName(GetCEPNameFromCCEP(cep, ns)); exists {
+		// add a cep into the ces
+		c.addCEPtoCES(cep, cesName)
+		return cesName
+	}
+
+	// If given cep object isn't packed in any of the CES. find a new ces
+	// to pack this cep.
+	cb := func() *cesTracker {
+		// get first available CES
+		for _, ces := range c.desiredCESs.getAllCESs() {
+			ces.backendMutex.RLock()
+			if ces.ces.Namespace != ns || len(ces.ces.Endpoints) >= c.maxCEPsInCES || len(ces.ces.Endpoints) == 0 {
+				ces.backendMutex.RUnlock()
+				continue
+			}
+			ces.backendMutex.RUnlock()
+			return ces
+		}
+		// allocate a new cesTracker and return
+		newCES := c.createCES("")
+		// Update the namespace to CES
+		newCES.ces.Namespace = ns
+		return newCES
+	}()
+
+	// Cache CEP name with newly allocated CES.
+	c.desiredCESs.insertCEP(GetCEPNameFromCCEP(cep, ns), cb.ces.GetName())
+
+	// Queue the CEP in CES
+	c.addCEPtoCES(cep, cb.ces.GetName())
+	return cb.ces.GetName()
+}
+
+// RemoveCEPFromCache is used to remove the CEP from local cache, this may result in
+// Updating an existing CES object.
+func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+	log.WithFields(logrus.Fields{
+		"CEPName": cepName,
+	}).Debug("Remove CEP from local cache")
+
+	// Check in local cache, if a given cep is already batched in one of the ces.
+	// and if exists, delete cep from ces.
+	if cesName, exists := c.desiredCESs.getCESName(cepName); exists {
+		var ces *cesTracker
+		if ces, exists = c.desiredCESs.getCESTracker(cesName); !exists {
+			log.Info("Valid CEP-CES mapping, but CES doesn't exist in CES cache.")
+			return
+		}
+
+		ces.backendMutex.Lock()
+		defer ces.backendMutex.Unlock()
+		for i, ep := range ces.ces.Endpoints {
+			if GetCEPNameFromCCEP(&ep, ces.ces.Namespace) == cepName {
+				// Insert deleted CoreCEP in removedCEPs
+				ces.removedCEPs[GetCEPNameFromCCEP(&ep, ces.ces.Namespace)] = struct{}{}
+				ces.ces.Endpoints =
+					append(ces.ces.Endpoints[:i],
+						ces.ces.Endpoints[i+1:]...)
+				break
+			}
+		}
+		log.WithFields(logrus.Fields{
+			"CESName":  cesName,
+			"CEPName":  cepName,
+			"CEPCount": len(ces.ces.Endpoints),
+		}).Debug("Removed CEP from CES")
+
+		// Increment the cepRemove counter
+		ces.cepRemoved += 1
+		c.insertCESInWorkQueue(ces, baseDelay)
+	} else {
+		log.WithFields(logrus.Fields{
+			logfields.CEPName: cepName,
+		}).Error("Failed to get CES from CESToCEPMapping")
```

Comment:
Simiarly here, we log an ERROR level msg, but it's not actionable. Why did it fail? What is the user supposed to takeaway from this msg? What are the consequences of failing to get the CES from the mapping in this function? These are the kinds of things that should be relayed to the user. If it's not necessary to relay this information, then consider downgrading the log level.

Reply:
Also, ensuring that the log msg itself has relevant logfields attached with information to troubleshoot. In this case, we have no idea which CES "failed".

Reply:
I hear your concerns, I'll downgrade this to `info` message. This information is critical to analyze if something went wrong in the cluster. Especially, if we have duplicate `CEP` delete requests coming into operator from api-server or In debugging why particular `CEP` delete request never propagated to cilium-agents.

>>Also, ensuring that the log msg itself has relevant logfields attached with information to troubleshoot.
The only information we have here is `CEP` object itself, we can't find matching `CES`.

Reply:
Ah that's helpful context actually. So this can happen if there's a duplicate delete of CEP? That might be worth throwing into the log msg, as a hint to suggest to developers / users what to look for.

Reply:
Thats correct, if there is duplicate `delete` request for CEP from `k8s-apiserver`, this would log error.  For `CEP` deletes `RemoveCEPFromCache` is the first functions gets called, so existing log message should be sufficient.

Reply:
Sure it might be sufficient, but I imagine that it's not obvious to any one looking at the logs who is not failure with the implementation details of this feature. That's why I think it's helpful to hint at it in the msg, otherwise this msg looks like a very generic error, but gives no clues as to why it happened.

Reply:
Looks like this feedback has been addressed. Chris is out on PTO now so I don't think it makes sense to wait on him to come back and double-check. We can always iterate further on items like logging over time as we get user feedback.